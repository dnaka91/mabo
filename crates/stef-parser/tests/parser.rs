use std::{
    fmt::{self, Display},
    fs,
};

use insta::{assert_snapshot, glob, with_settings};
use miette::{Diagnostic, MietteHandler, MietteHandlerOpts, ReportHandler};
use stef_parser::Schema;

#[test]
fn parse_schema() {
    glob!("inputs/*.stef", |path| {
        let input = fs::read_to_string(path).unwrap();
        let value = Schema::parse(input.as_str()).unwrap();

        with_settings!({
            description => input.trim(),
            omit_expression => true,
        }, {
            assert_snapshot!("parse", format!("{value:#?}"));
            assert_snapshot!("print", value.to_string());
        });
    });
}

#[test]
fn parse_invalid_schema() {
    struct Wrapper<'a>(&'a MietteHandler, &'a dyn Diagnostic);

    impl<'a> Display for Wrapper<'a> {
        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
            self.0.debug(self.1, f)
        }
    }

    let handler = MietteHandlerOpts::new()
        .color(false)
        .terminal_links(false)
        .width(120)
        .force_graphical(true)
        .build();

    glob!("inputs/invalid/*.stef", |path| {
        let input = fs::read_to_string(path).unwrap();
        let value = Schema::parse(input.as_str()).unwrap_err();

        with_settings!({
            description => input.trim(),
            omit_expression => true,
        }, {
            assert_snapshot!("error", Wrapper(&handler, &*value).to_string());
        });
    });
}
