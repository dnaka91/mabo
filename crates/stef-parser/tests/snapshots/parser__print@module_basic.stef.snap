---
source: crates/stef-parser/tests/parser.rs
description: "mod a {\n    /// Inner module\n    mod b {\n        enum Sample {\n            One @1,\n        }\n    }\n\n    struct Sample {\n        value: u32 @1,\n        inner: b::Sample @2,\n    }\n}"
input_file: crates/stef-parser/tests/inputs/module_basic.stef
---
mod a {
    /// Inner module
    mod b {
        enum Sample {
            One @1,
        }
    }

    struct Sample {
        value: u32 @1,
        inner: b::Sample @2,
    }
}


